autoscale/src/main/java/com/sequenceiq/periscope/service/MandatoryScalingAdjustmentService.java

```diff
@@ -0,0 +1,145 @@
+
+        loadAlert = cluster.getLoadAlerts().iterator().next();
+        loadAlertConfiguration = loadAlert.getLoadAlertConfiguration();
+        policyHostGroup = loadAlert.getScalingPolicy().getHostGroup();
+
+        hostFqdnsToInstanceId = stackResponseUtils.getCloudInstanceIdsForHostGroup(stackResponse, policyHostGroup);
+        servicesHealthyHostInstanceIds = stackResponseUtils.getCloudInstanceIdsWithServicesHealthyForHostGroup(stackResponse,
+                policyHostGroup);
+        stoppedHostInstanceIds = stackResponseUtils.getStoppedCloudInstanceIdsInHostGroup(stackResponse, policyHostGroup);
+
+        int stoppedNodeCount = stoppedHostInstanceIds.size();
+        int existingClusterNodeCount = stackResponse.getNodeCount() - stoppedNodeCount;
+
+        publishScalingEventIfNeeded(cluster, existingClusterNodeCount, pollingUserCrn, stackResponse, scalingAdjustmentParameters);
+    }
+
+    private void publishScalingEventIfNeeded(Cluster cluster, int existingClusterNodeCount, String pollingUserCrn, StackV4Response stackV4Response,
+            MandatoryScalingAdjustmentParameters scalingAdjustmentParameters) {
+
+        if (scalingAdjustmentParameters.getMandatoryStopStartUpscaleAdjustment().isPresent()) {
+            Integer targetScaleUpCount = scalingAdjustmentParameters.getMandatoryStopStartUpscaleAdjustment()
+                    .stream().map(adj -> IntStream.of(adj, loadAlertConfiguration.getMaxScaleUpStepSize()).min().getAsInt()).findFirst().orElse(0);
+            if (stoppedHostInstanceIds.isEmpty()) {
+                scalingEventSender.sendScaleUpEvent(loadAlert, existingClusterNodeCount, hostFqdnsToInstanceId.size(), servicesHealthyHostInstanceIds.size(),
+                        targetScaleUpCount);
+                LOGGER.info("Triggered mandatory adjustment ScaleUp for Cluster '{}', NodeCount '{}', HostGroup '{}'",
+                        cluster.getStackCrn(), targetScaleUpCount, policyHostGroup);
+            } else {
+                scalingEventSender.sendStopStartScaleUpEvent(loadAlert, existingClusterNodeCount, servicesHealthyHostInstanceIds.size(),
+                        stoppedHostInstanceIds.size());
+                LOGGER.info("Triggered mandatory adjustment stop-start ScaleUp for Cluster '{}', NodeCount '{}', HostGroup '{}'",
+                        cluster.getStackCrn(), stoppedHostInstanceIds.size(), policyHostGroup);
+            }
+        } else if (scalingAdjustmentParameters.getMandatoryStopStartDownscaleAdjustment().isPresent()) {
+            List<String> hostsToDecommission = collectDownscaleRecommendations(cluster, hostFqdnsToInstanceId, stoppedHostInstanceIds,
+                    pollingUserCrn, stackV4Response, scalingAdjustmentParameters.getMandatoryStopStartDownscaleAdjustment());
+
+            scalingEventSender.sendScaleDownEvent(loadAlert, hostFqdnsToInstanceId.size(), hostsToDecommission, servicesHealthyHostInstanceIds.size(), REGULAR);
```

Comment:
I guess here you should use `STOPSTART` instead of `REGULAR`

Reply:
These are mandatory downscale adjustments which remove a combination of stop nodes and yarn downscale recommendations. Since nodes are being removed and not stopped, the adjustmentType is `REGULAR`